Readme.md
# idb-batch

> Perform batch operation on IndexedDB

[![](https://saucelabs.com/browser-matrix/idb-batch.svg)](https://saucelabs.com/u/idb-batch)

[![](https://img.shields.io/npm/v/idb-batch.svg)](https://npmjs.org/package/idb-batch)
[![](https://img.shields.io/travis/treojs/idb-batch.svg)](https://travis-ci.org/treojs/idb-batch)
[![](http://img.shields.io/npm/dm/idb-batch.svg)](https://npmjs.org/package/idb-batch)

Create/update/remove objects from IndexedDB store in one transaction [without blocking the main thread](http://stackoverflow.com/questions/10471759/inserting-large-quantities-in-indexeddbs-objectstore-blocks-ui).
This module also manually validates unique indexes, fixing bugs in [WebKit](https://bugs.webkit.org/show_bug.cgi?id=149107)
and [IndexedDBShim](https://github.com/axemclion/IndexedDBShim/issues/56).

## Example

Using [idb-factory](https://github.com/treojs/idb-factory) and [ES2016 async/await syntax](https://jakearchibald.com/2014/es7-async-functions/).
Check [test.js](./test/index.js) for more examples.

```js
import batch from 'idb-batch'
import { open } from 'idb-factory'

// open IndexedDB database with 2 stores
const db = await open('mydb', 1, upgradeCallback)

// modify object store
await batch(db, 'magazines', [
  { type: 'add', key: 1, value: { name: 'M1', frequency: 12 } },
  { type: 'add', key: 2, value: { name: 'M2', frequency: 24 } },
  { type: 'add', key: 3, value: { name: 'M3', frequency: 6 } },
  { type: 'del', key: 4 },
]).then((result) => {
  console.log(result) // [1, 2, 3, undefined]
}).catch((err) => {
  console.error(err)
})

function upgradeCallback(e) {
  e.target.result.createObjectStore('books', { keyPath: 'id' })  
  e.target.result.createObjectStore('magazines')  
}
```

### transactionalBatch(tr: IDBDatabase|IDBTransaction, storeOpsArr, opts = { parallel: false, extraStores: [], resolveEarly: false })

`transactionalBatch` allows you to make operations across multiple stores
within a single transaction.

You may pass in your own `IDBTransaction` as `tr` or just supply an
`IDBDatabase` object and let `transactionalBatch` iterate over your operations
to determine what store names should be part of the transaction.

`storeOpsArr` is an array whose items are either functions (which accept the
transaction shared by the batch as argument) or objects whose keys are store
names and whose values follow the Array or Object notation described under
the `ops` argument of `batch`. `storeOpsArr` can also be such a single object
keyed by store name (its results will be presented in the resolved Promise
as if the object had been supplied within a (single-item) array).

If you are using store-name-keyed objects within `storeOpsArr`, be aware
that iteration order may not be consistent within an object, even when
`parallel` is false (see below), so you may wish to limit yourself to one
store key per object and use more objects.

`transactionalBatch` returns a `Promise` which resolves with the results of
each request. The results will be an array containing a child array for each
child of `storeOpsArr`. Within those child arrays will be, when functions are
supplied, the return result or, when objects are supplied, an object whose
keys are store names and whose values are the array of results for each child
operation.

#### `transactionalBatch` option: `extraStores`: Array

If you are using functions which operate on store names not specified elsewhere
among the (Object-based) operations, you may add those store names to the
`extraStores` property to allow the transaction shared with the functions to
operate on those store names.

#### `transactionalBatch` option: `parallel`: Boolean

The default behavior is to conduct operations in series, waiting for each
operation to complete before proceeding to the next (including waiting for
promises to resolve if a function-based operation supplies them as its own
return value, noting however that there is some risk in doing so, especially
when chaining promises, in that the transaction of the batch may actually
complete (and its promise thus resolve) before the operation's promise
resolves). If your batch of operations does not need to occur in order,
you may instead set the `parallel` option to `true` which will load operations
in parallel, and will not wait for operations to complete before proceeding
to the next, nor will it wait for all operations pertaining to a given store
or function to complete before proceeding to the next set of operations.

#### `transactionalBatch` option: `resolveEarly`: Boolean

By default, the promise returned by `transactionalBatch` will only resolve
with its array of results collecting return values for each of the
operations will only be made available after the transaction completes.
Setting `resolveEarly` to `true` will allow the promise to resolve with
the results array before it may be fully populated but after it has begun
execution of all of the operations within the batch.

#### `transactionalBatch` option: `adapterCb`: Function

By default, any function-type operations will be supplied the transaction
and the return result will be added to the results.

`adapterCb` is an optional callback which will be supplied the transaction
as well as the currently iterated function-type operation. The return
result of this callback will be added to the results. This allows for
supplying additional or alternative arguments to the callback (as opposed to
just the transaction object).

### batch(db: IDBDatabase|IDBTransaction, storeName: String, ops: Array|Object, opts: { parallel: false, extraStores: [], resolveEarly: false })

This creates a `readwrite` transaction to `storeName`, and performs `ops`
sequentially. It returns a `Promise` which resolves with the results of
each request.

This function only operates on a single store; if you require operations across
multiple stores, use `transactionalBatch`. See `transactionalBatch` also for
a description of the properties that may be used on `opts`.

Although this function allows an `IDBTransaction` in place of `IDBDatabase`,
usually, it may be generally more convenient to just pass in `IDBDatabase` and
let `batch` build the transaction for you (by auto-detecting the store names
implicit in `ops`).

**Array notation** is inspired by [LevelUP](https://github.com/Level/levelup#batch).
Each operation is an object with 3 possible properties: `type`, `key`, `value`.
`type` is either `add`, `put`, `del`, `move`, `copy`, or `clear`, and `key` is
optional when the store has a `keyPath` and the supplied value contains it.
When `move` or `copy` is used, the `key` represents the new destination of
the operation and `value` represents the old key from which to move or copy.

```js
await batch(db, 'books', [
  { type: 'add', key: 1, value: { name: 'M1', frequency: 12 } },
  { type: 'del', key: 2 },
  { type: 'put', value: { id: 3, name: 'M3', frequency: 24 } }, // no key
])
```

Note that `move` is composed of two operations, `copy` and `del`, and, at
present, does not function atomically, so if you set `parallel` to `true`,
there is a small chance that subsequent operations could occur after the
`copy` but before the `del` takes place (though the deleting will not take
place until a copy has been made).

At present, there are no options to allow `copy` or `move` to work across
stores or to involve sub-objects. You should instead use functions to conduct
such operations.

**Array notation (succinct)** is sugar for the array notation.

The above batch operation can be expressed in the following more succinct
format:

```js
await batch(db, 'books', [
  { add: { key: 1, value: { name: 'M1', frequency: 12 } } },
  { del: { key: 2 } },
  { put: { value: { id: 3, name: 'M3', frequency: 24 } } }, // no key
])
```

Multiple operations can also be merged into a single child array:

```js
await batch(db, 'books', [
  { add: [
    { key: 1, value: { name: 'M1', frequency: 12 } } }
    { key: 2, value: { name: 'M2', frequency: 8 } } }
  ] }
])
```

**Object notation** is sugar on top of array notation for `put`/`del`
operations. Set `key` to `'\0'` in order to delete a value.

```js
await batch(db, 'storage', {
  key1: 'update value',
  key2: '\0', // delete value
  key3: 'new value',
  key4: '\0\0abc' // To  get a single initial NUL, an extra one must be added
})
```

### getStoreNames(storeOps: Array|Object)

Returns an array of the store names used in a set of store operations.
This function does not obtain store names used solely within function
operations. If you wish for a function operation to operate on additional
store names and as part of the same transaction, you should add the required
store names to the `extraStores` option of `transactionalBatch`.

### ConstraintError

If during sequential execution one of the operations throws a
`ConstraintError`, the `Promise` rejects with an error, but previous
successful operations will commit.
This behavior may change in future versions, as I figure out how to
properly abort transactions in `IndexedDBShim`.

## LICENSE

[MIT](./LICENSE)


History.md
# idb-batch

## ?

* Breaking change: for object syntax, modify merge patch behavior to use '\0'
  for deletions and requiring strings with initial '\0' to include one extra;
  add tests
* Breaking change (minor): Switch to promise rejections instead of immediately synchronous errors for `batch()`/`transactionalBatch()`
* Breaking change (minor): Allow succincter `{add: {}}` and `{add: [{}]}` style operations
* Feature: Support `transactionalBatch`
* Feature: Added `getStoreNames`
* Feature: Allow preexisting transaction to be supplied to `batch()` (or `transactionalBatch()`)
* Feature: Support `move`, `copy` operations

## 1.0.0 / 2015-12-06

* initial release :sparkles:


test/index.js
// setup PhantomJS, Webkit, IE env
import 'polyfill-function-prototype-bind'
import 'indexeddbshim'
import ES6Promise from 'es6-promise'
import 'regenerator-runtime-only/runtime'

ES6Promise.polyfill()
if (navigator.userAgent.indexOf('Trident') !== -1) {
  console.log('force IE to enable compound indexes using indexeddbshim') // eslint-disable-line
  window.shimIndexedDB.__useShim()
}

import { expect } from 'chai'
import { del, open } from 'idb-factory'
import { request } from 'idb-request'
import Schema from 'idb-schema'
import batch, { transactionalBatch, getStoreNames } from '../src'

describe('idb-batch', function batchDesc() {
  this.timeout(5000)

  let db
  const dbName = 'idb-batch'
  const schema = new Schema()
  .addStore('books')
  .addIndex('byTitle', 'title', { unique: true }) // simple index
  .addIndex('byAuthor', 'author')
  .addStore('magazines', { key: 'id', increment: true })
  .addIndex('byName', 'name')
  .addIndex('byNameAndFrequency', ['name', 'frequency'], { unique: true }) // compound index
  .addStore('storage')
  .addIndex('byFoo', 'foo')

  beforeEach(async () => {
    db = await open(dbName, schema.version(), schema.callback())
  })

  before(() => del(db || dbName))
  afterEach(() => del(db || dbName))

  it('supports object syntax', async () => {
    const res1 = await batch(db, 'books', {
      key1: { title: 'B1', author: 'Bob' },
      key2: { title: 'B2', author: 'Bob' },
      3: { title: 'B3', author: 'Karl' },
    })

    expect(res1.sort()).eql(['key1', 'key2', '3'].sort()) // object keys don't guarantee order
    expect(await count(db, 'books')).equal(3)
    expect(await get(db, 'books', 'key1')).eql({ title: 'B1', author: 'Bob' })
    expect(await get(db, 'books', '3')).eql({ title: 'B3', author: 'Karl' })
    expect(await get(db, 'books', 3)).eql(undefined)

    const res2 = await batch(db, 'books', {
      key1: '\0',
      key2: '\0',
      3: { title: 'B3', author: 'Bob' },
    })

    expect(res2.sort()).eql([undefined, undefined, '3'].sort())
    expect(await count(db, 'books')).equal(1)
    expect(await get(db, 'books', '3')).eql({ title: 'B3', author: 'Bob' })
  })

  it('supports array syntax', async () => {
    const res1 = await batch(db, 'magazines', [
      { type: 'add', key: 1, value: { name: 'M1', frequency: 12 } },
      { type: 'add', key: 2, value: { name: 'M2', frequency: 24 } },
      { type: 'add', value: { id: 3, name: 'M3', frequency: 6 } },
      { type: 'add', value: { id: 4, name: 'M4', frequency: 52 } },
    ])

    expect(res1).eql([1, 2, 3, 4])
    expect(await count(db, 'magazines')).equal(4)
    expect(await get(db, 'magazines', 2)).eql({ id: 2, name: 'M2', frequency: 24 })
    expect(await get(db, 'magazines', 4)).eql({ id: 4, name: 'M4', frequency: 52 })

    const res2 = await batch(db, 'magazines', [
      { type: 'del', key: 1 },
      { type: 'put', key: 2, value: { name: 'M2', frequency: 24, foo: 'bar' } },
      { type: 'del', key: 3 },
    ])

    expect(res2).eql([undefined, 2, undefined])
    expect(await count(db, 'magazines')).equal(2)
    expect(await get(db, 'magazines', 2)).eql({ id: 2, name: 'M2', frequency: 24, foo: 'bar' })

    const res3 = await batch(db, 'magazines', [{ type: 'clear' }])
    expect(res3).eql([undefined])
    expect(await count(db, 'magazines')).equal(0)
  })

  it('supports special array syntax', async () => {
    const res1 = await batch(db, 'magazines', [
      { add: { key: 1, value: { name: 'M1', frequency: 12 } } },
      { add: [
        { key: 2, value: { name: 'M2', frequency: 24 } },
        { value: { id: 3, name: 'M3', frequency: 6 } },
        { value: { id: 4, name: 'M4', frequency: 52 } },
      ] },
    ])

    expect(res1).eql([1, 2, 3, 4])
    expect(await count(db, 'magazines')).equal(4)
    expect(await get(db, 'magazines', 2)).eql({ id: 2, name: 'M2', frequency: 24 })
    expect(await get(db, 'magazines', 4)).eql({ id: 4, name: 'M4', frequency: 52 })

    const res2 = await batch(db, 'magazines', [
      { del: { key: 1 } },
      { put: { key: 2, value: { name: 'M2', frequency: 24, foo: 'bar' } } },
      { del: [{ key: 3 }] },
    ])

    expect(res2).eql([undefined, 2, undefined])
    expect(await count(db, 'magazines')).equal(2)
    expect(await get(db, 'magazines', 2)).eql({ id: 2, name: 'M2', frequency: 24, foo: 'bar' })

    const res3 = await batch(db, 'magazines', [{ clear: {} }])
    expect(res3).eql([undefined])
    expect(await count(db, 'magazines')).equal(0)
  })

  it('works with any type of data (not only objects)', async () => {
    await batch(db, 'storage', {
      key1: 'value',
      key2: 123456,
      key3: [1, 2, 3],
      key4: { foo: 'bar' },
      key5: '\0abc',
      key6: '\0\0def',
    })

    expect(await count(db, 'storage')).equal(6)
    expect(await get(db, 'storage', 'key1')).equal('value')
    expect(await get(db, 'storage', 'key2')).equal(123456)
    expect(await get(db, 'storage', 'key3')).eql([1, 2, 3])
    expect(await get(db, 'storage', 'key4')).eql({ foo: 'bar' })
    expect(await get(db, 'storage', 'key5')).eql('abc')
    expect(await get(db, 'storage', 'key6')).eql('\0def')
  })

  it('validates unique indexes', async () => {
    const errors = []

    try {
      await batch(db, 'books', {
        1: { title: 'book', author: 'Petr' },
        2: { title: 'book', author: 'John' }, // error byTitle index
        3: { title: 'my book', author: 'John' },
      })
    } catch (err) {
      errors.push('simple index')
    }

    try {
      await batch(db, 'magazines', [
        { type: 'add', value: { name: 'magazine', frequency: 1 } },
        { type: 'add', value: { name: 'magazine', frequency: 1 } }, // error byNameAndFrequency index
        { type: 'add', value: { name: 'magazine', frequency: 2 } },
      ])
    } catch (err) {
      errors.push('compound index')
    }

    expect(errors).eql(['simple index', 'compound index'])
    expect(await count(db, 'books')).equal(1) // because transaction wasn't aborted
    expect(await count(db, 'magazines')).equal(1)
  })

  it('validates arguments', async (done) => {
    const errors = []
    const funcs = [
      async () => await batch(db, 'books'),
      async () => await batch(db, 123, { a: 1 }),
      async () => await batch(db, 'books', JSON.stringify({ a: 1 })),
      async () => await batch(db, 'magazines', [{ type: 'delete', key: 'foo' }]),
      async () => await batch(db, 'magazines', [['put', '1']]),
    ]
    const funcVals = funcs.values()

    const runFunc = async () => {
      const iter = funcVals.next()
      if (iter.done) {
        const errs = [
          'invalid arguments length',
          'invalid "storeName"',
          'invalid "ops"',
          'invalid type "delete"',
          'invalid op',
        ]
        expect(errors).eql(errs)
        done()
      }
      const func = iter.value
      try {
        await func()
      } catch (err) {
        errors.push(err.message)
      }
      runFunc()
    }
    runFunc()
  })

  describe('transactionalBatch', () => {
    it('supports batch in series', async () => {
      let gotCbResults = false
      let gotCb2Results = false
      let prom
      const res = await transactionalBatch(db, [
        {
          magazines: [
            { type: 'add', key: 1, value: { name: 'M1', frequency: 12 } },
            { type: 'add', key: 2, value: { name: 'M2', frequency: 24 } },
            { type: 'add', key: 3, value: { name: 'M3', frequency: 6 } },
            { type: 'del', key: 2 },
          ],
        },
        function callbackInTransaction(tr) {
          prom = new Promise((resolve) => {
            const magazines = tr.objectStore('magazines')
            const req = magazines.add({ name: 'M4', frequency: 8, [magazines.keyPath]: 5 })
            req.onsuccess = (e) => {
              expect(e.target.result).equal(5)
              // We can't do a timeout here as with the parallel test as we need the transaction to be the same
              const req2 = magazines.put({ name: 'M1', frequency: 17, [magazines.keyPath]: 1 })
              req2.onsuccess = (e2) => {
                expect(e2.target.result).equal(1)
                gotCbResults = true
                expect(gotCb2Results).equal(false)
                resolve('finished')
              }
            }
          })
          return prom
        },
        function callback2InTransaction(tr) {
          const magazines = tr.objectStore('magazines')
          const req = magazines.get(1)
          req.onsuccess = (e) => {
            gotCb2Results = true
            expect(e.target.result).eql({ name: 'M1', frequency: 17, id: 1 })
          }
        },
        {
          books: [
            { type: 'put', key: 1, value: { name: 'M1', frequency: 12 } },
            { type: 'move', key: 2, value: 1 },
            { type: 'copy', key: 3, value: 2 },
          ],
          storage: 'clear',
        },
      ])

      expect(res).eql([{ magazines: [1, 2, 3, undefined] }, prom, undefined, { books: [1, 2, undefined, 3], storage: [undefined] }])
      expect(await count(db, 'magazines')).equal(3)
      expect(await count(db, 'books')).equal(2)
      expect(await count(db, 'storage')).equal(0)
      expect(gotCbResults).equal(true)
    })

    it('supports batch in parallel', async (done) => {
      let gotCbResults = false
      let gotCb2Results = false
      const res = await transactionalBatch(db, [
        {
          magazines: [
            { type: 'add', key: 1, value: { name: 'M1', frequency: 12 } },
            { type: 'add', key: 2, value: { name: 'M2', frequency: 24 } },
            { type: 'add', key: 3, value: { name: 'M3', frequency: 6 } },
            { type: 'del', key: 2 },
          ],
        },
        function callbackInTransaction(tr) {
          return new Promise((resolve) => {
            const magazines = tr.objectStore('magazines')
            const req = magazines.add({ name: 'M4', frequency: 8, [magazines.keyPath]: 5 })
            req.onsuccess = (e) => {
              expect(e.target.result).equal(5)
              setTimeout(() => { // To test parallel we need a timeout, but this requires our needing to
                // create the transaction anew (though after a long enough time to ensure the original transaction is not still open)
                const trans = db.transaction('magazines', 'readwrite')
                const mags = trans.objectStore('magazines')
                const req2 = mags.put({ name: 'M1', frequency: 17, [mags.keyPath]: 1 })
                req2.onsuccess = (e2) => {
                  expect(e2.target.result).equal(1)
                  gotCbResults = true
                  expect(gotCb2Results).equal(true)
                  resolve('finished')
                }
              }, 500)
            }
          })
        },
        function callback2InTransaction(tr) {
          const magazines = tr.objectStore('magazines')
          const req = magazines.get(1)
          req.onsuccess = (e) => {
            gotCb2Results = true
            expect(gotCbResults).equal(false)
            expect(e.target.result).eql({ name: 'M1', frequency: 12, id: 1 })
          }
        },
        {
          books: [
            { type: 'put', key: 1, value: { name: 'M1', frequency: 12 } },
            { type: 'move', key: 2, value: 1 },
            { type: 'copy', key: 3, value: 2 },
          ],
          storage: 'clear',
        },
      ], { parallel: true })
      expect(res).eql([{ magazines: [1, 2, 3, undefined] }, new Promise(() => {}), undefined, { books: [1, 2, undefined, 3], storage: [undefined] }])
      expect(await count(db, 'magazines')).equal(3)
      expect(await count(db, 'books')).equal(2)
      expect(await count(db, 'storage')).equal(0)
      res[1].then((result) => {
        expect(result).equal('finished')
        done()
      })
    })

    it('supports resolving early (and continuing transaction)', async (done) => {
      const ops = [
        {
          magazines: [
            { type: 'add', key: 1, value: { name: 'M1', frequency: 12 } },
            { type: 'add', key: 2, value: { name: 'M2', frequency: 24 } },
            { type: 'add', key: 3, value: { name: 'M3', frequency: 6 } },
            { type: 'del', key: 2 },
          ],
        },
      ]
      const trans = db.transaction(getStoreNames(ops), 'readwrite')
      const res = await transactionalBatch(trans, ops, { resolveEarly: true })
      expect(res).eql([{ magazines: [1, 2, 3, undefined] }])
      const magazines = trans.objectStore('magazines')
      const req = magazines.add({ name: 'M4', frequency: 8, [magazines.keyPath]: 5 })
      req.onsuccess = (e) => {
        expect(e.target.result).equal(5)
        done()
      }
    })

    it('supports aborting a batch transaction', async () => {
      try {
        await transactionalBatch(db, [
          {
            magazines: [
              { type: 'add', key: 1, value: { name: 'M1', frequency: 12 } },
              { type: 'add', key: 2, value: { name: 'M2', frequency: 24 } },
              { type: 'add', key: 3, value: { name: 'M3', frequency: 6 } },
              { type: 'del', key: 2 },
            ],
          },
          function callbackInTransaction(tr) {
            tr.abort()
          },
        ])
      } catch (err) {
        expect(await count(db, 'magazines')).equal(0)
      }
    })
  })
})

/**
 * IDB store helpers.
 */

function count(db, storeName) {
  return request(db.transaction(storeName).objectStore(storeName).count())
}

function get(db, storeName, key) {
  return request(db.transaction(storeName).objectStore(storeName).get(key))
}


src/index.js
import 'babel-polyfill' // Array.prototype.values, etc.

import isPlainObj from 'is-plain-obj'
import isSafari from 'is-safari'

/**
 * Links to array prototype methods.
 */

const slice = [].slice
const map = [].map

/**
 * Perform batch operation for a single object store using `ops`.
 *
 * Array syntax:
 *
 * [
 *   { type: 'add', key: 'key1', val: 'val1' },
 *   { type: 'put', key: 'key2', val: 'val2' },
 *   { type: 'del', key: 'key3' },
 * ]
 *
 * Object syntax:
 *
 * {
 * 	 key1: 'val1', // put val1 to key1
 * 	 key2: 'val2', // put val2 to key2
 * 	 key3: null,   // delete key
 * }
 *
 * @param {Array|Object} ops Operations
 * @param {Object} opts See `transactionalBatch`
 * @return {Promise} Resolves to the result of the operations
 */

export default function batch(db, storeName, ops, opts) {
  if (typeof storeName !== 'string') return Promise.reject(new TypeError('invalid "storeName"'))
  if (![3, 4].includes(arguments.length)) return Promise.reject(new TypeError('invalid arguments length'))
  try {
    validateAndCanonicalizeOps(ops)
  } catch (err) {
    return Promise.reject(err)
  }
  return transactionalBatch(db, { [storeName]: ops }, opts).then((arr) => arr[0][storeName])
}

export function getStoreNames(storeOpsArr) {
  if (isPlainObj(storeOpsArr)) storeOpsArr = [storeOpsArr]
  return storeOpsArr.reduce((storeNames, opObj) => {
    // This has no effect if the opObj is a function
    return storeNames.concat(Object.keys(opObj))
  }, [])
}

/**
 * Perform batch operations for any number of object stores using `ops`.
 *
 * Array syntax:
 *
 * [
 *   { type: 'add', key: 'key1', val: 'val1' },
 *   { type: 'put', key: 'key2', val: 'val2' },
 *   { type: 'del', key: 'key3' },
 * ]
 *
 * Object syntax:
 *
 * {
 * 	 key1: 'val1', // put val1 to key1
 * 	 key2: 'val2', // put val2 to key2
 * 	 key3: null,   // delete key
 * }
 *
 * @param {IDBDatabase|IDBTransaction} tr IndexedDB database or transaction on which the batch will operate
 * @param {Array|Object} storeOpsArr Array of objects (or a single object) whose keys are store names and objects are idb-batch operations (object or array)
 * @param {Object} [opts] Options object
 * @property {Boolean} [opts.parallel=false] Whether or not to load in parallel
 * @property {Boolean} [opts.resolveEarly=false] Whether or not to resolve the promise before the transaction ends
 * @property {Array} [opts.extraStores=[]] A list of store names to add to the transaction (when `tr` is an `IDBDatabase` object)
 * @property {Function} [opts.adapterCb=null] A callback which will be supplied the `tr` and function of a function-type operation)
 * @return {Promise} Resolves to an array containing the results of the operations for each store
 */

export function transactionalBatch(tr, storeOpsArr, opts = { parallel: false, extraStores: [], resolveEarly: false, adapterCb: null }) {
  if (![2, 3].includes(arguments.length)) return Promise.reject(new TypeError('invalid arguments length'))
  if (isPlainObj(storeOpsArr)) storeOpsArr = [storeOpsArr]
  const storeOpsArrIter = storeOpsArr.keys()
  opts = opts || {}
  if (typeof tr.createObjectStore === 'function') {
    tr = tr.transaction(getStoreNames(storeOpsArr).concat(opts.extraStores || []), 'readwrite')
  }
  return new Promise((resolve, reject) => {
    const results = []
    tr.addEventListener('error', handleError(reject))
    tr.addEventListener('abort', handleError(reject))
    if (!opts.resolveEarly) tr.addEventListener('complete', () => resolve(results))

    let iterateStoreOps
    const iterateStores = (ct, storeResults, storeNames, storeOpsKeys, storeOpsObj, serial) => {
      if (typeof storeOpsObj === 'function') {
        let ret
        try {
          ret = opts.adapterCb ? opts.adapterCb(tr, storeOpsObj) : storeOpsObj(tr)
        } catch (err) {
          reject(err)
          return
        }
        results[ct] = ret
        if (serial) {
          if (ret && ret.then) {
            ret.then(iterateStoreOps)
            return
          }
          iterateStoreOps()
        }
        return
      }
      const storeOpIter = storeOpsKeys.next()
      if (storeOpIter.done) {
        if (serial) iterateStoreOps()
        return
      }
      const idx = storeOpIter.value
      const storeName = storeNames[idx]
      let moveCt = 0
      let successCt = 0

      let ops = storeOpsObj[storeName]
      try {
        ops = validateAndCanonicalizeOps(ops)
      } catch (err) {
        reject(err)
        return
      }

      const store = tr.objectStore(storeName)

      if (serial) {
        next(ct, 0)
      } else {
        for (let i = 0; i < ops.length; i++) {
          next(ct, i)
        }
        iterateStores(ct, storeResults, storeNames, storeOpsKeys, storeOpsObj)
      }

      function next(storeOpsIdx, opIndex) {
        const { type, key } = ops[opIndex]
        if (type === 'clear') {
          request(storeOpsIdx, opIndex, storeName, store.clear())
          return
        }
        if (type === 'del') {
          request(storeOpsIdx, opIndex, storeName, store.delete(key))
          return
        }
        const val = ops[opIndex].val || ops[opIndex].value
        if (['move', 'copy'].includes(type)) {
          const req = store.get(val)
          req.onerror = handleError(reject)
          req.onsuccess = (e) => {
            ops.splice(opIndex, 1, { type: 'put', key, value: e.target.result })
            if (type === 'move') {
              if (!serial) moveCt++
              ops.splice(opIndex + 1, 0, { type: 'del', key: val })
            }
            next(storeOpsIdx, opIndex)
            if (!serial && type === 'move') next(storeOpsIdx, opIndex + 1)
          }
          return
        }
        if (key && store.keyPath) val[store.keyPath] = key

        countUniqueIndexes(store, key, val, (err, uniqueRecordsCounter) => {
          if (err) return reject(err)

          // We don't abort transaction here (we just stop execution)
          // Browsers' implementations also don't abort, and instead just throw an error
          if (uniqueRecordsCounter) return reject(new Error('Unique index ConstraintError'))
          request(storeOpsIdx, opIndex, storeName, store.keyPath ? store[type](val) : store[type](val, key))
        })
      }

      function request(storeOpsIdx, opIndex, storeNm, req) {
        req.onerror = handleError(reject)
        req.onsuccess = (e) => {
          if (!storeResults[storeNm]) {
            storeResults[storeNm] = []
          }
          storeResults[storeNm][successCt++] = e.target.result
          if (successCt === ops.length - moveCt) {
            results[storeOpsIdx] = storeResults
            if (serial) iterateStores(ct, storeResults, storeNames, storeOpsKeys, storeOpsObj, serial)
          } else if (serial) {
            next(storeOpsIdx, ++opIndex)
          }
        }
      }
    }

    iterateStoreOps = opts.parallel ? () => {
      for (let storeOpsObj = storeOpsArrIter.next(); !storeOpsObj.done; storeOpsObj = storeOpsArrIter.next()) {
        const val = storeOpsObj.value
        storeOpsObj = storeOpsArr[val]
        const storeNames = Object.keys(storeOpsObj)
        const storeOpsKeys = storeNames.keys()
        iterateStores(val, {}, storeNames, storeOpsKeys, storeOpsObj)
      }
      if (opts.resolveEarly) resolve(results)
    } : () => {
      let storeOpsObj = storeOpsArrIter.next()
      if (storeOpsObj.done) {
        if (opts.resolveEarly) resolve(results)
        return
      }
      const val = storeOpsObj.value
      storeOpsObj = storeOpsArr[val]
      const storeNames = Object.keys(storeOpsObj)
      const storeOpsKeys = storeNames.keys()
      iterateStores(val, {}, storeNames, storeOpsKeys, storeOpsObj, true)
    }
    iterateStoreOps()
  })
}

/**
 * Validate unique index manually.
 *
 * Fixing:
 * - https://bugs.webkit.org/show_bug.cgi?id=149107
 * - https://github.com/axemclion/IndexedDBShim/issues/56
 *
 * @param {IDBStore} store
 * @param {Any} val
 * @param {Function} cb(err, uniqueRecordsCounter)
 */

function countUniqueIndexes(store, key, val, cb) {
  // rely on native support
  if (!isSafari && global.indexedDB !== global.shimIndexedDB) return cb()

  const indexes = slice.call(store.indexNames).map((indexName) => {
    const index = store.index(indexName)
    const indexVal = isCompound(index)
    ? map.call(index.keyPath, (indexKey) => val[indexKey]).filter((v) => v)
    : val[index.keyPath]

    return [index, indexVal]
  }).filter(([index, indexVal]) => {
    return index.unique && (isCompound(index) ? indexVal.length : indexVal)
  })

  if (!indexes.length) return cb()

  let totalRequestsCounter = indexes.length
  let uniqueRecordsCounter = 0

  indexes.forEach(([index, indexVal]) => {
    const req = index.getKey(indexVal) // get primaryKey to compare with updating value
    req.onerror = handleError(cb)
    req.onsuccess = (e) => {
      if (e.target.result && e.target.result !== key) uniqueRecordsCounter++
      totalRequestsCounter--
      if (totalRequestsCounter === 0) cb(null, uniqueRecordsCounter)
    }
  })
}

/**
 * Check if `index` is compound
 *
 * @param {IDBIndex} index
 * @return {Boolean}
 */

function isCompound(index) {
  return typeof index.keyPath !== 'string'
}

/**
 * Create error handler.
 *
 * @param {Function} cb
 * @return {Function}
 */

function handleError(cb) {
  return (e) => {
    // prevent global error throw https://bugzilla.mozilla.org/show_bug.cgi?id=872873
    if (typeof e.preventDefault === 'function') e.preventDefault()
    cb(e.target.error)
  }
}

/**
 * Validate operations and canonicalize them into an array.
 *
 * @param {Array|Object} ops
 * @return {Array} Canonicalized operations array
 */

function validateAndCanonicalizeOps(ops) {
  if (ops === 'clear') {
    ops = [{ type: 'clear' }]
  }
  if (!Array.isArray(ops) && !isPlainObj(ops)) {
    throw new TypeError('invalid "ops"')
  }
  if (isPlainObj(ops)) {
    ops = Object.keys(ops).map((key) => {
      return { key, value: typeof ops[key] === 'string' ? ops[key].replace(/^\0/, '') : ops[key], type: ops[key] === '\0' ? 'del' : 'put' }
    })
  }
  function checkOp(op) {
    if (['add', 'put', 'del', 'move', 'copy', 'clear'].indexOf(op.type) === -1) throw new TypeError(`invalid type "${op.type}"`)
  }
  ops.forEach((op, i) => {
    if (!isPlainObj(op)) throw new TypeError('invalid op')
    const opKeys = Object.keys(op)
    const type = opKeys[0]
    if (opKeys.length === 1 && type !== 'type') {
      op = op[type]
      const opers = Array.isArray(op) ? op : [op]
      ops.splice(i, 1, ...(opers.map((oper) => {
        Object.assign(oper, { type })
        checkOp(oper)
        return oper
      })))
      return
    }
    checkOp(op)
  })
  return ops
}
